Fixed BingChat to recent API changes.

gunpal5 opened the pull request:
Fixed BingChat to recent API changes. Now it is absolutely necessary to provide all of the Bing Cookies.
Added codes to ensure the conversation is authenticated.
Added retry mechanism on WebSocketException

this PR provides fixes to issues https://github.com/bsdayo/BingChat/issues/40 https://github.com/bsdayo/BingChat/issues/39 https://github.com/bsdayo/BingChat/issues/38 https://github.com/bsdayo/BingChat/issues/36

To use command line tool: 
1) Copy Bing Cookies using Cookies Editor extension and save to a file.
2) Set BING_COOKIES_FILE environment variable by using this command
`set [BING_COOKIES_FILE=[<path to cookies file>]]`
3)  dotnet bingchat

Commit:
Fixed BingChat to recent APIs changes. Now it is absolutely necessary to provide all the Bing Cookies.
Added codes to ensure the conversation is authenticated.
Added retry mechanism on WebSocketException

Commit:
Removed launchSettings.

neon-sunset commented:
Thank you for the pull request @gunpal5. However, it cannot be merged as is. There are a few points to address first:
- Remove irrelevant comments
- Use JsonSerializerContext instead of reflection-based serialization (`<T>`)
- Double-check if the change to parse the cookie file is necessary - 
- Fix syntax formatting errors
- Remove transient build/IDE files

Commit:
Removed irrelevant comments
Used JsonSerializerContext instead of reflection-based serialization
Removed transient IDE files

gunpal5 commented:
@neon-sunset I made necessary changes you asked.

Regarding cookie files, I am not sure which cookie Bing is using to authenticate the chat. only passing _U and/or KievRPSSecAuth are not enough. someone else might figure out the right cookies to pass as arguments in `BingChatClientOptions`

Commit:
Fixed some syntax formatting errors

gunpal5 commented:
@neon-sunset It's been like a month, people need it. Please review and merge

neon-sunset commented:
I understand that you would like to have this PR merged.  
But I can't do it because it still has issues and does not approach the problem in a way this library is designed - it reverse engineers original `BingChat` logic and requires to configure only what's actually necessary. Forcing the users to dump all of the cookies being the only way for it to work is a workaround. My suggestion would be inspecting the API interaction flow closer and also checking out what the libraries in other languages do, and then porting their fixes to this PR.

neon-sunset commented on .vs/BingChat/FileContentIndex/fabd545b-3a9e-4b0f-9c70-8c16fd4a4260.vsidx:
This file does not belong to the PR

neon-sunset commented on src/BingChat/BingChatClient.cs:
```diff
@@ -107,11 +115,15 @@ public async Task<BingChatConversation> CreateConversation()
             throw new BingChatException(message);
         }
 
+        var encryptedConversationSignature = rawResponse.Headers.GetValues("X-Sydney-EncryptedConversationSignature").FirstOrDefault();
```
If this doesn't work with `encryptedConversationSignature`, shouldn't it be `First()` instead of `FirstOrDefault()`?

neon-sunset commented on src/BingChat/BingChatConversation.cs:
```diff
@@ -18,35 +21,23 @@ namespace BingChat;
     {
         _request = new BingChatRequest(clientId, conversationId, conversationSignature, tone);
+        _encryptedConversationSignature = encryptedConversationSignature;
+        this._cookies = cookies;
```
```suggestion
        _cookies = cookies;
```

neon-sunset commented on src/BingChat/BingChatConversation.cs:
```diff
@@ -169,16 +167,54 @@ public async Task<string> AskAsync(string message, CancellationToken ct = defaul
+            CreateConnectionFactory(),
             HubProtocol,
-            HubEndpoint,
+            new UriEndPoint(new Uri(uri)),
```
You need to cache this property instead of creating it on each request.

neon-sunset commented on src/dotnet-bingchat/Utils.cs:
```diff
@@ -8,9 +8,12 @@ internal static class Utils
         return new BingChatClient(new BingChatClientOptions
         {
             CookieU = string.IsNullOrWhiteSpace(cookie) ? null : cookie,
+            CookieFilePath = string.IsNullOrEmpty(cookieFile)? null: cookieFile,
```
```suggestion
            CookieFilePath = string.IsNullOrEmpty(cookieFile) ? null : cookieFile,
```

neon-sunset commented on src/dotnet-bingchat/dotnet-bingchat.csproj:
Please revert all changes in this file.

neon-sunset commented:
I'll need to give it another look over some time later with a proper IDE with analyzers.

neon-sunset commented on src/BingChat/BingChatConversation.cs:
```diff
@@ -169,16 +167,54 @@ public async Task<string> AskAsync(string message, CancellationToken ct = defaul
+            }
+            catch (WebSocketException)
+            {
+                Thread.Sleep(100);
```
You must never use `Thread.Sleep(...)` in `async` methods or methods that can be called from async. The correct way to do it is to use `Task.Delay(...)`.
